### PR TITLE
[extractor/ant1newsgr] Add new extractor

### PR DESCRIPTION
**Pull request by zmousm**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This patch adds support for video extraction from ant1news.gr, a popular Greek content provider and established TV network.

* Add new IEs
  * Ant1NewsGrBaseIE: Base IE class
  * Ant1NewsGrWatchIE: Extract videos from TV VOD pages
  * Ant1NewsGrArticleIE: Extract videos from news articles
  * Ant1NewsGrEmbedIE: Extract iframe-embeddable ant1news.gr videos
* NB: There is a different platform at vod.antenna.gr, which is not covered here
* The Generic extractor can also be used to extract videos from news article pages (through `Ant1NewsGrEmbed._extract_urls`), however a specific IE is used to extract JSON-LD `@type='NewsArticle'` metadata
* Ant1NewsGrArticleIE defers to Ant1NewsGrEmbedIE, either as a playlist or a single video
* Ant1NewsGrWatchIE and Ant1NewsGrEmbedIE query an API endpoint to extract metadata, get the respective stream/source URLs and detect video formats
  * The endpoint HTTP path varies per IE
  * Ant1NewsGrEmbedIE first resolves any redirects for its' own URL, to derive the proper base URL for the API query

This patch was ported from ytdl-org/youtube-dl#30228 with the following changes/improvements:
* Parse HLS subtitles

### CODE REVIEW

**Comment by zmousm** [comment]
Rebase on master since #1983 was merged

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:30]: why? Instead just use `'_type': 'url_transparent'` on the urls returned by `_extract_urls`

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:9]:
```suggestion
import urllib.parse
```
to avoid any confusion. We have similarly named functions in utils/compat

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:83]: input URL's scheme should not be followed. Use `self.http_scheme()` (which should be deprecated imho) or just `https` instead

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:81]: I don't understand why URL has to be parsed separately. Just make the netloc a named group in regex. Or just hardcode it in the API path since the extractor is only dealing with a single domain

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:84]: Is downloading webpage just to get description worth it?

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:157]: extremely similar to the other PR. maybe this should be generalized into a common function. I'm sure I have seen similar patterns in other extractors too. 

Just noting down, not actually in scope of this PR

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:164]: it wouldn't match if it isn't suitable, no?

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:163]: maybe just do `self._proto_relative_url` in generic to avoid this complexity

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:175]: again, is this necessary? can't the domain just be hard-coded?

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:183]: why just timestamp?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:30]: I am afraid I do not understand your point.

These functions (de)serialize the parent_info kwargs provided to `Ant1NewsGrEmbedIE._extract_urls`, namely metadata from the embedding page.  The latter function is used both by the generic extractor and `Ant1NewsGrArticleIE`, which in addition extracts JSON-LD metadata. Since the particular functions are only used once they could be in-lined, however this code is generic and it looked cleaner this way. However I do not see how returning `{'_type': 'url_transparent'}` changes anything so that such metadata from the embedding page turn up and can be used for the embedded video?

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: Both `scheme` and `netloc` could be matched in the URL. However `urlparse` is more robust in general (not in this case) and I did this also for consistency with other code that prepares parameters for calling `_download_and_extract_api_data`. I can change and avoid it.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:80]: In the spirit of harvesting any available metadata, I would say it is 😊

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:153]: I will nevertheless inline `quot_re` as suggested elsewhere.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:160]: Actually this is not for granted:
`_URL_RE` simplifies the query-string matching part of `_VALID_URL`, since that part is not appropriate while extracting the URL from the iframe src attribute value (and also matching such a boundary would overly complicate the regex). The extracted URL should thus be checked subsequently against the full `_VALID_URL` regex, to ensure it also matches the query part.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:171]: In my experience the request to the embeddable player URL is typically redirected to a different domain (currently: www.antenna.gr -> nkv.antenna.gr) and the API endpoint is only valid under that domain. Hard-coding any of that seems risky to me, as it would make the extractor more fragile.

Unfortunately the HEAD request in this case is followed by a GET request to the new location -- I found no decent way to use the same method on the second request.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:79]: In my experience with this site I think it is risky to change the scheme to `https` (or user-preferred scheme) and assume the API request will query the expected (valid) endpoint. Also see the related comment below re: resolving the canonical base URL.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:179]: I can generalize this but, practically speaking, the fields from the embedding page which may complement API data are timestamp and description. The latter is not always related to the specific video (for example a page may embed multiple videos).

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: `_extract_urls` is a `classmethod`, so I am afraid it won't work after all (due to the `prefer_insecure` parameter lookup).

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:79]: I have now given up on using the scheme from input URL (and have thus simplified affected methods and regexes).

**Comment by zmousm** [comment]
Rebase on master to fix merge conflicts (since #1979 was merged)

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: I completely removed this block since
* `_proto_relative_url` is subsequently called in `playlist_from_matches`
* it is now also called in `Ant1NewsGrArticleIE._real_extract`

**Comment by zmousm** [comment]
When either #2297 or this is merged, `_scale_thumbnails_to_max_width` perhaps can be moved to `utils`.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: After giving up on scheme matching, I think I have now simplified `EMBED_RE` as much as possible.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: I think I finally got your point, and thus tried to switch from `(un)smuggle_url` to using `url_transparent` in the last commits (af08027 for some changes in `url_result` and `playlist_from_matches`, and bb12bda).
Is this what you meant?

A perhaps important difference with this approach is that the source of the metadata (parent) gets to dictate what fields it overrides in the subsequently resolved extraction.

Comment (pukkandan) [reply yt_dlp/extractor/ant1newsgr.py]: Yes

> A perhaps important difference with this approach is that the source of the metadata (parent) gets to dictate what fields it overrides in the subsequently resolved extraction.

Is that a problem?

Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: > Is that a problem?

Not that I can think of. It did work out in this case.

**Comment by zmousm** [comment]
Rebase on master to fix merge conflicts (since #2297 was merged)

**Comment by zmousm** [comment]
> When either #2297 or this is merged, `_scale_thumbnails_to_max_width` perhaps can be moved to `utils`.

What do you think about this @pukkandan ?

**Comment by pukkandan** [comment]
I've never seen any other extractor use that pattern. Are these 2 websites perhaps using the same underlying API?

**Comment by zmousm** [comment]
> I've never seen any other extractor use that pattern. Are these 2 websites perhaps using the same underlying API?

I can't tell, this is not somehow evident.

**Comment by pukkandan** [comment]
Let me think about it. For now, leave it in the extractor

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: After rebasing and fixing merge conflicts (c7b91ddea0781abf1cc4a9550e9dd4b0713e57fe was replaced by 311b6615d85d3530f2709c50e4223ff3b6b14361) I realized I still need a way to call `playlist_from_matches` and have it set `url_transparent` and override some fields. So I added `video_kwargs` to it and also added `url_result_or_playlist_from_matches`, a wrapper that either calls `url_result` or the former method.

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: After rebasing and fixing merge conflicts (c7b91ddea0781abf1cc4a9550e9dd4b0713e57fe was replaced by 311b6615d85d3530f2709c50e4223ff3b6b14361) I realized I still need a way to call `playlist_from_matches` and have it set `url_transparent` and override some fields. So I added `video_kwargs` to it and also added `url_result_or_playlist_from_matches`, a wrapper that either calls `url_result` or the former method.

Comment (pukkandan) [reply yt_dlp/extractor/ant1newsgr.py]: > After rebasing and fixing merge conflicts (c7b91dd was replaced by 311b661) I realized I still need a way to call playlist_from_matches and have it set url_transparent and override some fields. So I added video_kwargs to it

I intentionally did not add such a key because it does not make sense to do, as we discussed in the other PR. It makes sense to propagate the title only when there is a single entry in the page. If there are multiple embeds, we do not want all titles to end up being the same as the playlist

Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py]: I am also overriding the timestamp here.

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [inline yt_dlp/extractor/ant1newsgr.py:48]:
```suggestion
            'title': info.get('title'),
```
due to some recent changes, this is now more appropriate

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [reply yt_dlp/extractor/ant1newsgr.py:115]: In principle a news article page may or may not embed video(s) in this way. Should an error be raised in the latter case (like so), since the extractor has already matched the URL? Or should it return the URL and defer to other extractors?
The reason for this extractor over the generic extractor is the extraction of JSON-LD metadata.

File: yt_dlp/extractor/common.py
Comment (zmousm) [reply yt_dlp/extractor/common.py:1142]: The first use case for this is in `Ant1NewsGrArticleIE` so as to override timestamp (with url_transparent).
I also use it in the generic extractor to override title, but I do not remember the use case for that anymore so I will skip that.
Besides that I hope you agree on the usefulness of this facility, per the [previous comment](https://github.com/yt-dlp/yt-dlp/pull/1982#discussion_r789909826).

File: yt_dlp/extractor/ant1newsgr.py
Comment (zmousm) [inline yt_dlp/extractor/ant1newsgr.py:102]: You removed this function in the last commit you pushed before merging.

**Review by zmousm — COMMENTED** [review]
Hey @pukkandan, thanks for merging! I noticed one issue after the fact, see inline.

File: yt_dlp/extractor/ant1newsgr.py
Comment (pukkandan) [reply yt_dlp/extractor/ant1newsgr.py:102]: my bad, will fix

**Pull request merged**
